### PR TITLE
tests: added template + metadata functional test

### PR DESCRIPTION
**Pull request by kayrus**
Covers https://github.com/coreos/fleet/issues/1446

### CODE REVIEW

**Comment by kayrus** [comment]
@jonboulle reverting of PR #1376 resolved the issue. We can keep these tests and merge this PR. Or I can spend more time and fix initial PR. What is your decision?

**Comment by jonboulle** [comment]
#1376 did not make it into a release right? If not, let's revert that + add the test for this release, and then we can rework a proper version for the next release. Thanks!

**Comment by kayrus** [comment]
@jonboulle no, it didn't. tests cover the bug. new #1514 issue has been created. waiting for LGTM ;)

File: functional/metadata_test.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+	_, err = cluster.WaitForNActiveUnits(m0, 3)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ndesired := 3
+	if stdout, stderr, err := cluster.Fleetctl(m0, "list-units", "--no-legend", "--full", "--fields", "unit,active,machine"); err != nil {
+		t.Fatalf("Unable to get submited units: \nstdout: %s\nstderr: %s\nerr: %v", stdout, stderr, err)
+	} else {
```
Comment (jonboulle) [inline functional/metadata_test.go]: Remove the else - Fatal will exit anyway

File: functional/metadata_test.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	}
+
+	_, err = cluster.WaitForNActiveUnits(m0, 3)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ndesired := 3
+	if stdout, stderr, err := cluster.Fleetctl(m0, "list-units", "--no-legend", "--full", "--fields", "unit,active,machine"); err != nil {
+		t.Fatalf("Unable to get submited units: \nstdout: %s\nstderr: %s\nerr: %v", stdout, stderr, err)
```
Comment (jonboulle) [inline functional/metadata_test.go]: submitted

File: functional/metadata_test.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+			}
+			currentMachine := fmt.Sprintf("smoke%s", state.Machine)
+			if desiredMachine[1] != currentMachine {
+				t.Fatalf("Template (%s) has been scheduled on wrong machine (%s): \nstdout: %s\nstderr: %s", state.Name, currentMachine, stdout, stderr)
+			}
+		}
+	}
+
+	if stdout, stderr, err := cluster.Fleetctl(m0, "start", "--block-attempts=20", "fixtures/units/metadata@invalid.service"); err == nil {
+		t.Fatalf("metadata@invalid unit should not be scheduled: \nstdout: %s\nstderr: %s\nerr: %v", stdout, stderr, err)
```
Comment (jonboulle) [inline functional/metadata_test.go]: Printing err here isn't exactly telling you much ;-)

File: functional/metadata_test.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ndesired := 3
+	if stdout, stderr, err := cluster.Fleetctl(m0, "list-units", "--no-legend", "--full", "--fields", "unit,active,machine"); err != nil {
+		t.Fatalf("Unable to get submited units: \nstdout: %s\nstderr: %s\nerr: %v", stdout, stderr, err)
+	} else {
+		stdout = strings.TrimSpace(stdout)
+		if err != nil {
```
Comment (jonboulle) [inline functional/metadata_test.go]: wah? this slipped in here. You checked err on https://github.com/coreos/fleet/pull/1512/files#diff-0f4946ffffcab3f77e657cfe24b2e8f5R63

**Comment by kayrus** [comment]
@jonboulle Done.

**Comment by tixxdz** [comment]
@jonboulle I discussed this with @kayrus, so maybe spend some time on it? cause it's really an optimization boost, previous logic triggered unnecessary RPC and I/O work. The code that can be fixed was located, and there is already a functional test for it, otherwise I'm ok.

**Comment by antrik** [comment]
@tixxdz I think your last comment should rather go into the newly opened issue, https://github.com/coreos/fleet/issues/1514 ? I was quite confused what it refers to at first...

**Comment by jonboulle** [comment]
I think you're right.
@kayrus thanks, this is great!

**Comment by antrik** [comment]
@kayrus some formal nitpicks: first of all, please update the title of this PR -- it doesn't apply anymore, and it's very confusing...

I think the new tests should be applied *after* the revert, to preserve bisectability.

Also, this might be a matter of taste -- but I believe it would be better to have only one revert commit covering both of the original commits (i.e. using `revert -n`); and with a proper commit message explaining why they are reverted...

**Comment by antrik** [comment]
Meh, too late ;-)

**Comment by jonboulle** [comment]
Sorry, all fair comments, just wanted to unblock.

**Pull request merged**